File: src/test/java/tech/aroma/data/sql/serializers/MessageSerializerTest.java
package tech.aroma.data.sql.serializers; /**
 * @author SirWellington
 */

import java.sql.Timestamp;
import java.time.Duration;
import java.time.Instant;
import java.util.*;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.*;
import org.springframework.jdbc.core.JdbcTemplate;
import tech.aroma.thrift.Message;
import tech.sirwellington.alchemy.test.junit.runners.*;

import static org.hamcrest.Matchers.*;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.*;
import static tech.sirwellington.alchemy.arguments.Arguments.*;
import static tech.sirwellington.alchemy.arguments.assertions.TimeAssertions.equalToInstantWithinDelta;
import static tech.sirwellington.alchemy.arguments.assertions.TimeAssertions.inTheFuture;
import static tech.sirwellington.alchemy.generator.AlchemyGenerator.one;
import static tech.sirwellington.alchemy.generator.NumberGenerators.integers;
import static tech.sirwellington.alchemy.generator.StringGenerators.alphabeticString;
import static tech.sirwellington.alchemy.test.junit.ThrowableAssertion.*;

@RunWith(AlchemyTestRunner.class)
@Repeat(100)
public class MessageSerializerTest
{
    @Mock
    private JdbcTemplate database;

    @GenerateString
    private String statement;

    @GeneratePojo
    private Message message;

    @GenerateString(GenerateString.Type.UUID)
    private String appId;

    @GenerateString(GenerateString.Type.UUID)
    private String messageId;

    @GenerateString(GenerateString.Type.ALPHABETIC)
    private String alphabetic;

    private MessageSerializer instance;

    @Captor
    private ArgumentCaptor<Object> captor;

    @Before
    public void setup()
    {
        message.messageId = messageId;
        message.applicationId = appId;

        instance = new MessageSerializer();
    }

    @Test
    public void testSave() throws Exception
    {
        instance.save(message, null, statement, database);
        checkMessageWithDuration(message, null);
    }

    @Test
    public void testSaveWithExpiration() throws Exception
    {
        int daysToLive = one(integers(3, 10));
        Duration ttl = Duration.ofDays(daysToLive);

        instance.save(message, ttl, statement, database);

        checkMessageWithDuration(message, ttl);
    }

    @DontRepeat
    @Test
    public void testWhenDatabaseFails() throws Exception
    {
        when(database.update(anyString(), Matchers.<Object>anyVararg()))
                .thenThrow(new RuntimeException());

        assertThrows(() -> instance.save(message, null, statement, database));
    }

    @DontRepeat
    @Test
    public void testSaveWithInvalidIDs() throws Exception
    {
        message.messageId = alphabetic;

        assertThrows(() -> instance.save(message, null, statement, database))
                .isInstanceOf(IllegalArgumentException.class);
    }

    @DontRepeat
    @Test
    public void testSaveWithBadArgs() throws Exception
    {
        String statement = one(alphabeticString());

        assertThrows(() -> instance.save(null, null, statement, database))
                .isInstanceOf(IllegalArgumentException.class);

        assertThrows(() -> instance.save(message, null, null, database))
                .isInstanceOf(IllegalArgumentException.class);

        assertThrows(() -> instance.save(message, null, "", database))
                .isInstanceOf(IllegalArgumentException.class);

        assertThrows(() -> instance.save(message, null, statement, null))
                .isInstanceOf(IllegalArgumentException.class);

        message.applicationId = statement;
        message.messageId = statement;
        assertThrows(() -> instance.save(message, null, statement, database))
                .isInstanceOf(IllegalArgumentException.class);
    }

    private void checkMessageWithDuration(Message message, Duration ttl)
    {
        verify(database).update(eq(statement), captor.capture());

        List<Object> arguments = captor.getAllValues();

        assertThat(arguments.get(0), is(UUID.fromString(messageId)));
        assertThat(arguments.get(1), is(message.title));
        assertThat(arguments.get(2), is(message.body));
        assertThat(arguments.get(3), is(message.urgency.toString()));
        assertThat(arguments.get(4), is(epochToTimestamp(message.timeOfCreation)));
        assertThat(arguments.get(5), is(epochToTimestamp(message.timeMessageReceived)));

        Object expiration = arguments.get(6);
        if (Objects.nonNull(ttl))
        {
            checkExpirationWithTTL(expiration, ttl);
        }

        assertThat(arguments.get(7), is(message.hostname));
        assertThat(arguments.get(8), is(message.macAddress));
        assertThat(arguments.get(9), is(UUID.fromString(appId)));
        assertThat(arguments.get(10), is(message.applicationName));
        assertThat(arguments.get(11), is(message.deviceName));
    }

    private void checkExpirationWithTTL(Object expiration, Duration ttl)
    {
        assertThat(expiration, notNullValue());
        assertThat(expiration, instanceOf(Timestamp.class));

        Instant actualExpiration = ((Timestamp) expiration).toInstant();
        Instant expectedExpiration = Instant.now().plus(ttl);

        long acceptableDelta = 100;

        checkThat(actualExpiration)
                .is(inTheFuture())
                .is(equalToInstantWithinDelta(expectedExpiration, acceptableDelta));
    }

    public Timestamp epochToTimestamp(long epochMilli)
    {
        Instant instant = Instant.ofEpochMilli(epochMilli);
        return Timestamp.from(instant);
    }
}

File: src/test/java/tech/aroma/data/sql/serializers/MessageSerializerTest.kt
package tech.aroma.data.sql.serializers

import com.nhaarman.mockito_kotlin.eq
import com.nhaarman.mockito_kotlin.whenever
import org.hamcrest.Matchers
import org.junit.Assert.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentCaptor
import org.mockito.Captor
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.Mockito.verify
import org.springframework.jdbc.core.JdbcTemplate
import tech.aroma.data.sql.toTimestamp
import tech.aroma.data.sql.toUUID
import tech.aroma.thrift.Message
import tech.sirwellington.alchemy.generator.StringGenerators.Companion.alphabeticStrings
import tech.sirwellington.alchemy.generator.one
import tech.sirwellington.alchemy.test.junit.ThrowableAssertion.assertThrows
import tech.sirwellington.alchemy.test.junit.runners.AlchemyTestRunner
import tech.sirwellington.alchemy.test.junit.runners.DontRepeat
import tech.sirwellington.alchemy.test.junit.runners.GeneratePojo
import tech.sirwellington.alchemy.test.junit.runners.GenerateString
import tech.sirwellington.alchemy.test.junit.runners.Repeat
import java.sql.ResultSet
import org.hamcrest.Matchers.`is` as Is

/**
 * @author SirWellington
 */

@RunWith(AlchemyTestRunner::class)
@Repeat(100)
class MessageSerializerTest
{
    @Mock
    private lateinit var database: JdbcTemplate

    @Mock
    private lateinit var resultSet: ResultSet

    @GenerateString
    private lateinit var sql: String

    @GeneratePojo
    private lateinit var message: Message

    @GenerateString(GenerateString.Type.UUID)
    private lateinit var appId: String

    @GenerateString(GenerateString.Type.UUID)
    private lateinit var messageId: String

    @GenerateString(GenerateString.Type.ALPHABETIC)
    private lateinit var alphabetic: String

    private lateinit var instance: MessageSerializer

    @Captor
    private lateinit var captor: ArgumentCaptor<Any>

    @Before
    fun setup()
    {
        message.messageId = messageId
        message.applicationId = appId

        message.unsetIsTruncated()
        setupResults()

        instance = MessageSerializer()
    }

    @Test
    @Throws(Exception::class)
    fun testSave()
    {
        instance.save(message, sql, database)

        verify(database).update(sql,
                                messageId.toUUID(),
                                appId.toUUID(),
                                message.applicationName,
                                message.title,
                                message.body,
                                message.urgency.toString(),
                                message.timeOfCreation.toTimestamp(),
                                message.timeMessageReceived.toTimestamp(),
                                message.hostname,
                                message.macAddress,
                                message.deviceName)
    }


    @DontRepeat
    @Test
    @Throws(Exception::class)
    fun testWhenDatabaseFails()
    {
        whenever(database.update(eq(sql), Mockito.anyVararg<Any>()))
                .thenThrow(RuntimeException())

        assertThrows { instance.save(message, sql, database) }
    }

    @DontRepeat
    @Test
    @Throws(Exception::class)
    fun testSaveWithInvalidIDs()
    {
        message.messageId = alphabetic

        assertThrows {
            instance.save(message, sql, database)
        }.isInstanceOf(IllegalArgumentException::class.java)
    }

    @DontRepeat
    @Test
    @Throws(Exception::class)
    fun testSaveWithBadArgs()
    {
        val statement = one(alphabeticStrings())

        assertThrows {
            instance.save(message, "", database)
        }.isInstanceOf(IllegalArgumentException::class.java)

        message.applicationId = statement
        message.messageId = statement

        assertThrows {
            instance.save(message, statement, database)
        }.isInstanceOf(IllegalArgumentException::class.java)
    }

    @Test
    fun testDeserialize()
    {
        val result = instance.deserialize(resultSet)

        assertThat(result, Matchers.equalTo(message))
    }

    private fun setupResults()
    {
        whenever(resultSet.getString(Columns.Messages.MESSAGE_ID)).thenReturn(message.messageId)
        whenever(resultSet.getString(Columns.Messages.APP_ID)).thenReturn(message.applicationId)
        whenever(resultSet.getString(Columns.Messages.APP_NAME)).thenReturn(message.applicationName)
        whenever(resultSet.getString(Columns.Messages.TITLE)).thenReturn(message.title)
        whenever(resultSet.getString(Columns.Messages.BODY)).thenReturn(message.body)
        whenever(resultSet.getString(Columns.Messages.HOSTNAME)).thenReturn(message.hostname)
        whenever(resultSet.getString(Columns.Messages.IP_ADDRESS)).thenReturn(message.macAddress)
        whenever(resultSet.getString(Columns.Messages.DEVICE_NAME)).thenReturn(message.deviceName)
        whenever(resultSet.getString(Columns.Messages.PRIORITY)).thenReturn(message.urgency.toString())

        whenever(resultSet.getTimestamp(Columns.Messages.TIME_CREATED)).thenReturn(message.timeOfCreation.toTimestamp())
        whenever(resultSet.getTimestamp(Columns.Messages.TIME_RECEIVED)).thenReturn(message.timeMessageReceived.toTimestamp())
    }

}